// vant/lib/action-sheet/ActionSheet.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _vue = require("vue");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _utils = require("../utils");

var _icon = require("../icon");

var _popup = require("../popup");

var _loading = require("../loading");

var _shared = require("../popup/shared");

// Utils
// Components
var [name, bem] = (0, _utils.createNamespace)('action-sheet');

var _default = (0, _vue.defineComponent)({
  name,
  props: (0, _extends2.default)({}, _shared.popupSharedProps, {
    title: String,
    actions: Array,
    cancelText: String,
    description: String,
    closeOnPopstate: Boolean,
    closeOnClickAction: Boolean,
    round: {
      type: Boolean,
      default: true
    },
    closeable: {
      type: Boolean,
      default: true
    },
    closeIcon: {
      type: String,
      default: 'cross'
    },
    safeAreaInsetBottom: {
      type: Boolean,
      default: true
    }
  }),
  emits: ['select', 'cancel', 'update:show'],

  setup(props, {
    slots,
    emit
  }) {
    var updateShow = show => emit('update:show', show);

    var onCancel = () => {
      updateShow(false);
      emit('cancel');
    };

    var renderHeader = () => {
      if (props.title) {
        return (0, _vue.createVNode)("div", {
          "class": bem('header')
        }, [props.title, props.closeable && (0, _vue.createVNode)(_icon.Icon, {
          "name": props.closeIcon,
          "class": bem('close'),
          "onClick": onCancel
        }, null)]);
      }
    };

    var renderCancel = () => {
      if (slots.cancel || props.cancelText) {
        return [(0, _vue.createVNode)("div", {
          "class": bem('gap')
        }, null), (0, _vue.createVNode)("button", {
          "type": "button",
          "class": bem('cancel'),
          "onClick": onCancel
        }, [slots.cancel ? slots.cancel() : props.cancelText])];
      }
    };

    var renderOption = (item, index) => {
      var {
        name,
        color,
        subname,
        loading,
        callback,
        disabled,
        className
      } = item;
      var Content = loading ? (0, _vue.createVNode)(_loading.Loading, {
        "class": bem('loading-icon')
      }, null) : [(0, _vue.createVNode)("span", {
        "class": bem('name')
      }, {
        default: () => [name]
      }), subname && (0, _vue.createVNode)("div", {
        "class": bem('subname')
      }, {
        default: () => [subname]
      })];

      var onClick = () => {
        if (disabled || loading) {
          return;
        }

        if (callback) {
          callback(item);
        }

        if (props.closeOnClickAction) {
          updateShow(false);
        }

        (0, _vue.nextTick)(() => emit('select', item, index));
      };

      return (0, _vue.createVNode)("button", {
        "type": "button",
        "style": {
          color
        },
        "class": [bem('item', {
          loading,
          disabled
        }), className],
        "onClick": onClick
      }, {
        default: () => [Content]
      });
    };

    var renderDescription = () => {
      if (props.description || slots.description) {
        var content = slots.description ? slots.description() : props.description;
        return (0, _vue.createVNode)("div", {
          "class": bem('description')
        }, {
          default: () => [content]
        });
      }
    };

    var renderOptions = () => {
      if (props.actions) {
        return props.actions.map(renderOption);
      }
    };

    return () => (0, _vue.createVNode)(_popup.Popup, (0, _vue.mergeProps)({
      "class": bem(),
      "round": props.round,
      "position": "bottom",
      "safeAreaInsetBottom": props.safeAreaInsetBottom
    }, (0, _extends2.default)({}, (0, _utils.pick)(props, _shared.popupSharedPropKeys), {
      'onUpdate:show': updateShow
    })), {
      default: () => [renderHeader(), renderDescription(), (0, _vue.createVNode)("div", {
        "class": bem('content')
      }, [renderOptions(), slots.default == null ? void 0 : slots.default()]), renderCancel()]
    });
  }

});

exports.default = _default;
